# 02. Build SQLite Database (CoffeeKing)

**Goal:** Create a SQLite databae from Yelp JSON files so we can run SQL queries (joins, aggregations) for Milestone 3-4.

**Tables (initial):**
- `coffee_business` (from business.json, coffee-related subset only)
- `coffee_reviews` (from review.json, only reviews for coffee_businesss)

> Note: Raw JSON files are not committed to GitHub. The SQLite DB file is also not committed.

In [1]:
import pandas as pd
import sqlite3
from pathlib import Path

DATA_RAW = Path("../data_raw")
DB_PATH = Path("../db/coffeeking.db")

biz_path = DATA_RAW / "yelp_academic_dataset_business.json"
review_path = DATA_RAW / "yelp_academic_dataset_review.json"

conn = sqlite3.connect(DB_PATH)
print("DB connected:", DB_PATH.resolve())

DB connected: /Users/kwaknakyung/projects/Coffeeking-Yelp/db/coffeeking.db


## 2. Build `coffee_business` table (from business.json)

We filter coffee-related businesses using a simple keyword match on `categories`.
This keeps the DB light and focused for SQL analysis.

In [2]:
biz = pd.read_json(biz_path, lines= True)

biz["categories_clean"] = biz["categories"].fillna("").str.lower()
coffee_keywords = ["coffee", "cafe", "cafes", "coffee & tea"]

coffee_mask = biz["categories_clean"].str.contains("|".join(coffee_keywords), regex=True)
coffee_biz = biz.loc[coffee_mask].copy()

coffee_biz.shape

(8509, 15)

In [3]:
coffee_business = coffee_biz [
    ["business_id", "name", "address", "city", "state", "postal_code",
    "latitude", "longitude", "stars", "review_count", "is_open", "categories"]
].copy()

coffee_business.head(3)

business_id                             name  \
3   MTSW4McQd7CbVtyjqoe9mw               St Honore Pastries   
20  WKMJwqnfZKsAae75RMP6jA   Roast Coffeehouse and Wine Bar   
46  JX4tUpd09YFchLBuI43lGw  Naked Cyber Cafe & Espresso Bar   

                 address          city state postal_code   latitude  \
3            935 Race St  Philadelphia    PA       19107  39.955505   
20   10359 104 Street NW      Edmonton    AB     T5J 1B9  53.546045   
46  10303  108 Street NW      Edmonton    AB     T5J 1L7  53.544682   

     longitude  stars  review_count  is_open  \
3   -75.155564    4.0            80        1   
20 -113.499169    4.0            40        0   
46 -113.506589    4.0            12        1   

                                           categories  
3   Restaurants, Food, Bubble Tea, Coffee & Tea, B...  
20  Coffee & Tea, Food, Cafes, Bars, Wine Bars, Re...  
46  Arts & Entertainment, Music Venues, Internet S...

In [4]:
coffee_business.to_sql("coffee_business", conn, if_exists="replace", index=False)
print("Saved table: coffee_business")

Saved table: coffee_business


In [5]:
pd.read_sql_query("SELECT COUNT(*) AS n_rows FROM coffee_business;", conn)

n_rows
0    8509

## 3. Build `coffee_reviews` table (from review.json, filtered)

`review.json` is very large, so we read it in chunks.
We keep only reviews whose `business_id` exists in `coffee_business`.

In [6]:
# pull business_ids directly from SQLite (so DB is the source of truth)
coffee_ids_df = pd.read_sql_query("SELECT business_id FROM coffee_business;", conn)
coffee_ids = set(coffee_ids_df["business_id"])

len(coffee_ids)

8509

In [7]:
chunk_size = 50000
max_chunks = 2

kept = 0
chunks_done = 0

# start fresh for reproducibility
conn.execute("DROP TABLE IF EXISTS coffee_reviews;")
conn.commit()

for chunk in pd.read_json(review_path, lines=True, chunksize=chunk_size):
    # filter to only coffee business_ids
    chunk = chunk[chunk["business_id"].isin(coffee_ids)]

    if not chunk.empty:
        chunk_small = chunk[["review_id", "business_id", "user_id", "stars", "date", "text"]].copy()
        chunk_small.to_sql("coffee_reviews", conn, if_exists="append", index = False)
        kept += len(chunk_small)

    chunks_done += 1
    print(f"chunk {chunks_done} done | kept so far: {kept}")

    if chunks_done >= max_chunks:
        break

kept

chunk 1 done | kept so far: 5670
chunk 2 done | kept so far: 11351


11351

In [8]:
pd.read_sql_query("SELECT COUNT(*) AS n_rows FROM coffee_reviews;", conn)

n_rows
0   11351

## 4. Sanity check: join coffee_business + coffee_reviews

Goal: Verify the join works and that reviews aggregate correctly by geography.

In [10]:
q = """
SELECT
    b.state,
    COUNT(DISTINCT b.business_id) AS n_business,
    COUNT(r.review_id) AS n_reviews,
    ROUND(AVG(r.stars), 3) AS avg_review_stars
FROM coffee_business b
LEFT JOIN coffee_reviews r
ON b.business_id = r.business_id
GROUP BY b.state
ORDER BY n_reviews DESC
LIMIT 10;
"""
pd.read_sql_query(q, conn)

state  n_business  n_reviews  avg_review_stars
0    LA         633       2554             4.116
1    PA        2128       2084             3.918
2    FL        1519       1749             4.053
3    TN         647       1046             4.016
4    IN         573        910             4.057
5    MO         551        681             3.783
6    CA         243        587             4.083
7    NV         327        556             4.144
8    NJ         487        389             3.866
9    AZ         435        374             3.684

In [11]:
q2 = """
SELECT
    SUM(CASE WHEN r.n_reviews IS NULL THEN 1 ELSE 0 END) AS business_with_0_reviews,
    COUNT(*) AS total_businesses
FROM(
    SELECT business_id FROM coffee_business
) b
LEFT JOIN (
    SELECT business_id, COUNT(*) AS n_reviews
    FROM coffee_reviews
    GROUP BY business_id
) r
ON b.business_id = r.business_id;
"""
pd.read_sql_query(q2, conn)

business_with_0_reviews  total_businesses
0                     7876              8509

### Quick takeaways (sanity check)
- Join works: `coffee_business` <=> `coffee_reviews` by `business_id`.
- Current review coverage is very low because we only loaded 2 chunks of `review.json`.
    - Businesses with 0 loaded reviews: **7,876 / 8,509** (92.6%)
- Next: fully load `coffee_reviews` from `review.json` so review-based insights are representative.


## 5. Full load `coffee_reviews` (all matching reviews)

We rebuild `coffee_reviews`from scratch and load all reviews for coffee-related businesses.
This is required before any reliable review-based statistics or text analysis.

In [12]:
import time
import pandas as pd

review_path = "../data_raw/yelp_academic_dataset_review.json"
chunk_size = 50000

# coffee_ids must exist
# coffee_ids = set(coffee_biz["business_id"])

conn.execute("DROP TABLE IF EXISTS coffee_reviews;")
conn.commit()

kept = 0
chunks_done = 0
t0 = time.time()

for chunk in pd.read_json(review_path, lines=True, chunksize=chunk_size):
    chunk = chunk[chunk["business_id"].isin(coffee_ids)]
    if not chunk.empty:
        chunk_small = chunk[["review_id", "business_id", "user_id", "stars", "date", "text"]].copy()
        chunk_small.to_sql("coffee_reviews", conn, if_exists = "append", index = False)
        kept += len(chunk_small)

    chunks_done += 1
    if chunks_done % 10 == 0:
        elapsed = time.time() - t0
        print(f"chunks: {chunks_done} | kept: {kept} | elapsed: {elapsed:.1f}s")

kept

chunks: 10 | kept: 55502 | elapsed: 4.1s
chunks: 20 | kept: 102670 | elapsed: 8.2s
chunks: 30 | kept: 146450 | elapsed: 12.1s
chunks: 40 | kept: 189507 | elapsed: 16.0s
chunks: 50 | kept: 230068 | elapsed: 19.9s
chunks: 60 | kept: 271705 | elapsed: 23.8s
chunks: 70 | kept: 315770 | elapsed: 29.7s
chunks: 80 | kept: 358319 | elapsed: 35.2s
chunks: 90 | kept: 398562 | elapsed: 39.8s
chunks: 100 | kept: 442231 | elapsed: 44.2s
chunks: 110 | kept: 487771 | elapsed: 48.8s
chunks: 120 | kept: 535104 | elapsed: 53.1s
chunks: 130 | kept: 584129 | elapsed: 57.4s
chunks: 140 | kept: 630349 | elapsed: 61.7s


630349

In [13]:
pd.read_sql_query("SELECT COUNT(*) AS n_rows FROM coffee_reviews;", conn)

n_rows
0  630349

In [14]:
q = """
SELECT
    SUM(CASE WHEN r.n_reviews IS NULL THEN 1 ELSE 0 END) AS business_with_0_reviews,
    COUNT(*) AS total_businesses
FROM coffee_business b
LEFT JOIN(
    SELECT business_id, COUNT(*) AS n_reviews
    FROM coffee_reviews
    GROUP BY business_id
) r
ON b.business_id = r.business_id;
"""
pd.read_sql_query(q, conn)

business_with_0_reviews  total_businesses
0                        0              8509

## 6. Correlations (ratings vs popularity)

Now that all coffee-related reviews are loaded, we can study relationship such as:

- Are higher-rated coffee businesses also more "popular" (have more reviews)?
- Do some states/cities have systematically higher ratings or higher review volume?

We will compute:
1) Business-level average review stars from `coffee_reviews`
2) Correlation between review volume and average rating

In [15]:
q = """
SELECT
    b.business_id,
    b.name,
    b.city,
    b.state,
    b.stars AS business_stars, --Yelp business table rating
    b.review_count AS business_review_count, --Yelp business table count
    COUNT(r.review_id) AS n_reviews_loaded, --from review table
    AVG(r.stars) AS avg_review_stars
FROM coffee_business b
JOIN coffee_reviews r
ON b.business_id = r.business_id
GROUP BY b.business_id, b.name, b.city, b.state, b.stars, b.review_count;
"""
biz_summary = pd.read_sql_query(q, conn)
biz_summary.head()

business_id                             name          city  \
0  -0__F9fnKt8uioCKztF5Ww                         Piscobar   New Orleans   
1  -0eUa8TsXFFy0FCxHYmrjg   Waterfront Gourmet Cafe & Deli  Philadelphia   
2  -0epFLgYq2C1Jo_W4FOBKw                   Our House Cafe      Columbia   
3  -0gRYq5UjMtZbELj0KHxzA                La Colombe Coffee     Bryn Mawr   
4  -0iIxySkp97WNlwK66OGWg  Truckee Bagel Company - Midtown          Reno   

  state  business_stars  business_review_count  n_reviews_loaded  \
0    LA             4.0                     66                66   
1    PA             4.0                     26                28   
2    IL             4.0                     15                18   
3    PA             3.5                    128               133   
4    NV             3.5                    219               233   

   avg_review_stars  
0          3.878788  
1          3.964286  
2          4.222222  
3          3.593985  
4          3.721030

In [16]:
biz_summary.shape, biz_summary[["n_reviews_loaded", "avg_review_stars"]].describe()

((8509, 8),
        n_reviews_loaded  avg_review_stars
 count       8509.000000       8509.000000
 mean          74.080268          3.592373
 std          174.002102          0.951368
 min            5.000000          1.000000
 25%           12.000000          3.000000
 50%           27.000000          3.833333
 75%           69.000000          4.339286
 max         5778.000000          5.000000)

In [17]:
corr = biz_summary[["n_reviews_loaded", "avg_review_stars"]].corr(numeric_only=True)
corr

n_reviews_loaded  avg_review_stars
n_reviews_loaded          1.000000          0.153095
avg_review_stars          0.153095          1.000000

### Interpretation (Correlation)

- Correlation between **n_reviews_loaded** (popularity proxy) and **avg_review_stars** (quality proxy) is **0.153**.
- This is a **weak positive relationship**, meaning: businesses with more reviews are only *slightly* more likely to have average rating.
- Takeaway for CoffeeKing: popularity and quality should be treated as **separate signals** (don't assume "busy" = "best")

## 7. New Metric: Visibility Score (Popularity x Quality)

Review counts are extremely skewed (a few places have thousands of reviews), so we log-transform review volume.
Then we combine it with average review stars to capture "well-known AND well-liked" businesses.

**Visibility Score = log(1 + n_reviews_loaded) x avg_review_stars**

Why this helps CoffeeKing:
- Helps identify benchmark businessses that are both popular and consistently rated well.
- Produces a ranking that is less distorted by extreme outliers than raw review counts. 

In [18]:
import numpy as np

biz_summary["visibility_score"] = np.log1p(biz_summary["n_reviews_loaded"]) * biz_summary["avg_review_stars"]

biz_summary.sort_values("visibility_score", ascending=False)[
    ["name", "city", "state", "n_reviews_loaded", "avg_review_stars", "visibility_score"]
].head(15)

name              city state  n_reviews_loaded  \
8356       Reading Terminal Market      Philadelphia    PA              5778   
6903    Ruby Slipper - New Orleans       New Orleans    LA              5264   
482       Mazzaro's Italian Market  Saint Petersburg    FL              1583   
5725                        Bodega  Saint Petersburg    FL              1812   
7872     Surrey's Café & Juice Bar       New Orleans    LA              2120   
2849                 Cafe La Maude      Philadelphia    PA              1530   
1384  District Donuts Sliders Brew       New Orleans    LA              2110   
6208                          Parc      Philadelphia    PA              2884   
6045                    Willa Jean       New Orleans    LA              2050   
7095               Oxford Exchange             Tampa    FL              1938   
1749                Jimmy J's Cafe       New Orleans    LA              2178   
6095                     Baja Cafe            Tucson    AZ              1130   
3490             Ruby Slipper Cafe       New Orleans    LA              2176   
4377                Mr. B's Bistro       New Orleans    LA              2116   
3514     Handlebar Coffee Roasters     Santa Barbara    CA              1065   

      avg_review_stars  visibility_score  
8356          4.605400         39.891908  
6903          4.291223         36.770791  
482           4.641188         34.194918  
5725          4.556291         34.184662  
7872          4.460849         34.168511  
2849          4.633333         33.979367  
1384          4.378199         33.514751  
6208          4.146671         33.037692  
6045          4.246829         32.386671  
7095          4.270898         32.330388  
1749          4.202479         32.302867  
6095          4.585841         32.242392  
3490          4.183364         32.152093  
4377          4.192817         32.107564  
3514          4.571831         31.873291

In [19]:
(biz_summary["business_review_count"] - biz_summary["n_reviews_loaded"]).describe()

count    8509.000000
mean       -2.404513
std         5.193207
min      -123.000000
25%        -3.000000
50%        -1.000000
75%         0.000000
max         0.000000
dtype: float64

### Key Findings: "Visibility Score" Top Businesses

Top-ranked businesses by **Visibility Score = log(1+n_reviews_loaded) x avg_review_stars** are heavily concentrated in a few cities/states:

- **Philadelphia, PA** and **New Orleans, LA** appear repeatedly in the top list.
- **Florida (Saint Petersburg, Tampa)** also shows multiple high-visibility coffee-related destinations.

Interpretation:
- The most "visible" coffee-related listings often look like **coffee + destination food** hybrids (markets, brunch, donuts, etc.), not only small standalone coffee shops.
- This suggests CoffeeKing should separate strategy into:
(1) "destination/ food-forward café" positioning vs.
(2) "pure coffee shop" positioning,
because the success drivers (review_volume and customer expectations) may differ.

### Data sanity check: business review_count vs loaded reviews

We compared `busienss_review_count` (from business.json) with `n_reviews_loaded` (counted from review.json).

Result (business_review_count - n_reviews_loaded):
- Median difference is around **-1** and **75% of businesses have a difference of 0**.
- The minimum difference reaches **-123**, meaning a small number of businesses have slightly more reviews in review.json than the business.json review_count filed.

Takeaway:
- Overall, our review loading/join looks consistent.
- Small mismatches are expected in large public datasets and do not impact high-level insights.

## 8. Concept Tagging (Positioning) - Pure Coffee vs Food-forward

Goal: Create simple, explainable "concept tags" from the Yelp `categories` field so we can compare business performance by positioning.
This makes the project more actionable for CoffeeKing: it's not only "where to open" but also "what concept mix tends to win."

In [21]:
# Build concept tags in pandas (safe + explainable rules)

import re
import numpy as np

# normalize categories text
coffee_biz["categories_clean"] = coffee_biz["categories"].fillna("").str.lower()

# define keyword buckets (simple on purpose)
pure_coffee_kw = r"(coffee|coffee\s*&\s*tea|cafe|cafes|espresso|tea|bubble tea)"
food_kw = r"(breakfast|brunch|bakery|bakeries|donut|donuts|dessert|ice cream|sandwich|deli|restaurant|restaurants|pizza|burgers|barbeque|bbq|steak|tacos)"
alcohol_kw = r"(bar|bars|wine|beer|cocktail|brewpub|pub)"
market_kw = r"(market|grocery|supermarket|department stores|shopping)"

# tag presence flags
coffee_biz["has_food"] = coffee_biz["categories_clean"].str.contains(food_kw, regex = True)
coffee_biz["has_alcohol"] = coffee_biz["categories_clean"].str.contains(alcohol_kw, regex = True)
coffee_biz["has_market"] = coffee_biz["categories_clean"].str.contains(market_kw, regex = True)

# concept group (mutually exclusive label)
def concept_group(row):
    if row["has_market"]:
        return "market/retail"
    if row["has_alcohol"]:
        return "coffee+alcohol"
    if row["has_food"]:
        return "food-forward"
    return "pure-coffee"

coffee_biz["concept_group"] = coffee_biz.apply(concept_group, axis=1)
coffee_biz["concept_group"].value_counts()

/var/folders/dm/y0wy_ncs3qn20cs4xdn9sfq00000gn/T/ipykernel_37345/3766344450.py:16: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  coffee_biz["has_food"] = coffee_biz["categories_clean"].str.contains(food_kw, regex = True)
/var/folders/dm/y0wy_ncs3qn20cs4xdn9sfq00000gn/T/ipykernel_37345/3766344450.py:17: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  coffee_biz["has_alcohol"] = coffee_biz["categories_clean"].str.contains(alcohol_kw, regex = True)
/var/folders/dm/y0wy_ncs3qn20cs4xdn9sfq00000gn/T/ipykernel_37345/3766344450.py:18: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  coffee_biz["has_market"] = coffee_biz["categories_clean"].str.contains(market_kw, regex = True)


concept_group
food-forward      5530
pure-coffee       1363
coffee+alcohol    1076
market/retail      540
Name: count, dtype: int64

In [22]:
# Save enriched table to SQLite for SQL-first analysis

# keep only useful columns so the DB stays clean
coffee_business_enriched = coffee_biz[[
    "business_id", "name", "address", "city", "state", "postal_code",
    "latitude", "longitude", "stars", "review_count", "is_open", "categories",
    "concept_group", "has_food", "has_alcohol", "has_market"
]].copy()

coffee_business_enriched.to_sql("coffee_business_enriched", conn, if_exists = "replace", index = False)

# sanity check
pd.read_sql_query("""
SELECT concept_group, COUNT(*) AS n_business
FROM coffee_business_enriched
GROUP BY concept_group
ORDER BY n_business DESC;
""", conn)

concept_group  n_business
0    food-forward        5530
1     pure-coffee        1363
2  coffee+alcohol        1076
3   market/retail         540

## Concept performance - Do some concept mixes earn higher ratings and visuability?

Now that we tagged each coffee-related business into a simple positioning group, we compare outcomes by concept:
- average review stars (from `coffee_reviews`)
- review volume (n_reviews_loaded)
- visibility_score (log-scaled mix of volume + rating)
This helps CoffeeKing decide not only *where* to open, but *what kind of coffee concept* is more likely to succeed.

In [25]:
# SQL: compare concept groups (ratings, volume, visibility)

concept_perf = pd.read_sql_query("""
WITH concept_base AS (
  SELECT 
    b.business_id,
    b.concept_group,
    b.state,
    b.city
  FROM coffee_business_enriched b
),
review_agg AS (
  SELECT 
    business_id,
    COUNT(*) AS n_reviews_loaded,
    AVG(stars) AS avg_review_stars
  FROM coffee_reviews
  GROUP BY business_id
),
scored AS (
  SELECT
    cb.concept_group,
    ra.business_id,
    ra.n_reviews_loaded,
    ra.avg_review_stars,
    (LOG(1 + ra.n_reviews_loaded) * ra.avg_review_stars) AS visibility_score
  FROM review_agg ra
  JOIN concept_base cb
    ON ra.business_id = cb.business_id
)
SELECT
  concept_group,
  COUNT(*) AS n_business,
  ROUND(AVG(avg_review_stars), 3) AS avg_rating,
  ROUND(AVG(n_reviews_loaded), 1) AS avg_reviews,
  ROUND(AVG(visibility_score), 2) AS avg_visibility
FROM scored
GROUP BY concept_group
ORDER BY avg_visibility DESC;
""", conn)

concept_perf


concept_group  n_business  avg_rating  avg_reviews  avg_visibility
0  coffee+alcohol        1076       4.007        115.5            6.87
1   market/retail         540       4.059         77.2            5.88
2    food-forward        5530       3.470         75.1            5.44
3     pure-coffee        1363       3.575         36.1            4.85

## "Winner rate" - which concept is over-represented among top-visibility businesses?

Instead of only looking at averages (which can be skewed by huge outliers),
we measure how often each concept appears in the **top 10% visiblity** businesses.
This is a simple, business-friendly KPI: "What concepts show up disproportionately among the winners?"

In [26]:
# SQL: top 10% visiblity share by concept (winner rate)

winner_rate = pd.read_sql_query("""
WITH scored AS(
    SELECT
        b.concept_group,
        r.business_id,
        COUNT(*) AS n_reviews_loaded,
        AVG(r.stars) AS avg_review_stars,
        (LOG(1+COUNT(*)) * AVG(r.stars)) AS visibility_score
    FROM coffee_reviews r
    JOIN coffee_business_enriched b
    ON r.business_id = b.business_id
    GROUP BY b.concept_group, r.business_id
),
cutoff AS (
    SELECT
        visibility_score,
        NTILE(10) OVER (ORDER BY visibility_score DESC) AS decile
    FROM scored
),
scored_with_decile AS (
    SELECT
        s.*,
        NTILE(10) OVER (ORDER BY s.visibility_score DESC) AS decile
    FROM scored s
)
SELECT
    concept_group,
    COUNT(*) AS n_business,
    SUM(CASE WHEN decile = 1 THEN 1 ELSE 0 END) AS n_top10pct,
    ROUND(1.0 * SUM(CASE WHEN decile = 1 THEN 1 ELSE 0 END) / COUNT(*), 3) AS top10pct_rate
FROM scored_with_decile
GROUP BY concept_group
ORDER BY top10pct_rate DESC;
""", conn)

winner_rate

concept_group  n_business  n_top10pct  top10pct_rate
0  coffee+alcohol        1076         189          0.176
1   market/retail         540          59          0.109
2    food-forward        5530         560          0.101
3     pure-coffee        1363          43          0.032

### Interpretation (Concept performance)

- Overall, **coffee+alcohol** stands out as the strongest "visibility" concept:
    - highest avg_visibility (6.87)
    - high avg_rating (4.007) and the highest avg_reviews (115.5)
    - also dominates the "winner rate" metric: **17.6%** of coffee+alcohol businesses land in the top 10% visibility.

- **market/retail** is the most consistently high-rated group (avg_rating 4.059) with solid visibility (5.88), suggesting this concept may perform well when paired with a strong retail/market destination angle.

- **food-forward** is the largest bucket (5,530 businesses) but has the lowest avg_rating (3.470).
Its winner rate (10.1%) is close to market/retail (10.9), which suggests a small subset of food-forward "destination" places can still break out, but on average this concept is more competitive / mixed-quality.

- **pure-coffee** is the weakest on visibility:
    - lowest avg_reviews (36.1) and lowest avg_visibility (4.85)
    - winner rate is only **3.2%**, meaning it is under-represented among top performers. 

CoffeeKing takeaway:
If CoffeeKing's goal is to maximize early discoverability and demand, the data suggests prioritizing a **coffee+alcohol** hybrid (or a concept that boosts "evening + social" usage), instead of launching as a pure coffee shop. A safer "premium" alternative is **market/retail**, which shows strong ratings with decent visibility. 

### Location + Concept "Winner Rate" (CoffeeKing Recommendation Engine)

Goal:
Identify the best **city/state + concept_group** combinations for CoffeeKing by measuring "winner rate".

Definition:
- We already defined a **top performer** as businesses in the **top 10% of visibility_score**.
- Here, we compute winner rate by location and concept:
winner_rate = (# top10% businesses) / (total businesses)

Output:
Top 10 location + concept combos with:
- enough sample size (to avoid tiny cities dominating)
- high winner_rate
- plus helpful context (avg_rating, avg_reviews, avg_visibility)